libsignal-service/src/sender.rs

```diff
@@ -192,157 +208,264 @@ where
+            ContentBody::DataMessage(message) => {
+                message.flags == Some(Flags::EndSession as u32)
+            }
+            _ => false,
         };
 
-        for _ in 0..4 {
-            match self
-                .send_messages(
-                    recipient,
-                    unidentified_access.as_ref(),
-                    &content,
+        let result = self
+            .try_send_message(
+                recipient.clone(),
+                unidentified_access,
+                &content_body,
+                timestamp,
+                online,
+            )
+            .await;
+
+        match result {
+            Ok(SendMessageResult { needs_sync, .. }) if needs_sync => {
+                log::debug!("sending multi-device sync message");
+                let sync_message = self
+                    .create_multi_device_sent_transcript_content(
+                        Some(&recipient),
+                        Some(message),
+                        timestamp,
+                    )?;
+                self.try_send_message(
+                    (&self.cipher.local_address).clone(),
+                    None,
+                    &sync_message,
                     timestamp,
-                    online,
+                    false,
                 )
-                .await
-            {
-                Err(MessageSenderError::TryAgain) => continue,
-                r => return r,
+                .await?;
             }
+            _ => (),
         }
-        Err(MessageSenderError::MaximumRetriesLimitExceeded)
+
+        if end_session {
+            log::debug!("ending session with {}", recipient);
+            if let Some(ref uuid) = recipient.uuid {
+                self.cipher.store_context.delete_all_sessions(&uuid)?;
+            }
+            if let Some(ref e164) = recipient.e164 {
+                self.cipher.store_context.delete_all_sessions(&e164)?;
+            }
+        }
+
+        result
     }
 
-    async fn send_messages(
+    /// Send a message to the recipients in a group.
+    pub async fn send_message_to_group(
         &mut self,
-        recipient: &ServiceAddress,
-        _unidentified_access: Option<&UnidentifiedAccess>,
-        content: &[u8],
+        recipients: Vec<ServiceAddress>,
+        unidentified_access: Option<&UnidentifiedAccess>,
+        message: crate::proto::DataMessage,
         timestamp: u64,
         online: bool,
-    ) -> Result<SendMessageResponse, MessageSenderError> {
-        let messages = self
-            .create_encrypted_messages(&recipient, None, &content)
-            .await?;
-
-        let messages = OutgoingPushMessages {
-            destination: recipient.identifier(),
-            timestamp,
-            messages,
-            online,
-        };
+    ) -> Result<
+        Vec<Result<SendMessageResult, MessageSenderError>>,
+        MessageSenderError,
+    > {
+        // TODO: discuss and fix the rather horrible type above
```

Comment:
I think this is okay, to be honest. Folding it to `Result<Vec<SendMessageResult>, MessageSenderError>` is probably too loose.

Reply:
Why not `Vec<Result<SendMessageResult, MessageSenderError>>`? Because when resync is needed, just a retry of sending the message is done right? If this fails again, just replace the initial error with the new one?
Then the `Vec` contains one result for each message.

Reply:
> Why not `Vec<Result<SendMessageResult, MessageSenderError>>`? Because when resync is needed, just a retry of sending the message is done right? If this fails again, just replace the initial error with the new one?

Good point, I'll do exactly that.